README.md
# Simple Choghadiya Calculator (SCC)

* scc.py
Primary routine that handles the Choghadiya Calculation.
  [Read detailed documentation](docs/scc-help.md) about how to use this script.

* scc-cli.py
Command Line Interface for above mentioned scc.py. Can be used in other bash scripts. Output of this script can be parsed by command line parsers easily. Therefore, this script can be used in pipes as well.
  [Read detailed documentation](docs/scc-cli-help.md) about how to use this script.

* scc-gui.py
Platform independent Graphical User Interface to scc.py developed using Python's tKinter / ttk. Hence, this GUI application can be used on Windows, Linux or macOS. 
[Read detailed documentation](docs/scc-gui-help.md) about how to use this script.

docs/release_notes.md
# Release Notes
1. Now huge dictionary from scc.py has been replaces with a Generator, and a Choghadiya class that generates Choghadiya on the fly. This should reduce memory requirements of the script as well as automatic calculation reduces the chances of data-entry mistakes.

src/scc.py
#!/usr/bin/env python3
# -*- coding: utf_8 -*-
"""

    Copyright    : 2015 May. A. R. Bhatt.
    Organization : VAU SoftTech
    Project      : SCC
    Script Name  : scc.py
    License      : GNU General Public License v3.0


    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""

# All import statements for Gujarati પંચાંગ - ચોઘડિયા
import sys
from datetime import datetime, date, time, timedelta

# Constants for the names of the Gujarati પંચાંગ - ચોઘડિયા
CHOGHADIYU_AM: str = 'Amrut'
CHOGHADIYU_CH: str = 'Chal'
CHOGHADIYU_KA: str = 'Kaal'
CHOGHADIYU_LA: str = 'Laabh'
CHOGHADIYU_RO: str = 'Rog'
CHOGHADIYU_SH: str = 'Shubh'
CHOGHADIYU_UD: str = 'Udveg'

#
# Dictionary holding names of all the choghadiya for all the days of a week
# Structure of this nested dictionary is as follows:
# Out most dictionary holds key-value pair of weekday number and respective
# Choghadiya names for day and night as inner dictionary.
# Weekday number corresponds to weekday number returned by Python's Weekday
# method of the datetime class.
#
CHOGHADIYA_DICT = {
    0:  # MONDAY
        {'D': {1: CHOGHADIYU_AM,
               2: CHOGHADIYU_KA,
               3: CHOGHADIYU_SH,
               4: CHOGHADIYU_RO,
               5: CHOGHADIYU_UD,
               6: CHOGHADIYU_CH,
               7: CHOGHADIYU_LA,
               8: CHOGHADIYU_AM},
         'N': {1: CHOGHADIYU_CH,
               2: CHOGHADIYU_RO,
               3: CHOGHADIYU_KA,
               4: CHOGHADIYU_LA,
               5: CHOGHADIYU_UD,
               6: CHOGHADIYU_SH,
               7: CHOGHADIYU_AM,
               8: CHOGHADIYU_CH}
         },
    1:  # TUESDAY
        {'D': {1: CHOGHADIYU_RO,
               2: CHOGHADIYU_UD,
               3: CHOGHADIYU_CH,
               4: CHOGHADIYU_LA,
               5: CHOGHADIYU_AM,
               6: CHOGHADIYU_KA,
               7: CHOGHADIYU_SH,
               8: CHOGHADIYU_RO},
         'N': {1: CHOGHADIYU_KA,
               2: CHOGHADIYU_LA,
               3: CHOGHADIYU_UD,
               4: CHOGHADIYU_SH,
               5: CHOGHADIYU_AM,
               6: CHOGHADIYU_CH,
               7: CHOGHADIYU_RO,
               8: CHOGHADIYU_KA}
         },
    2:  # WEDNESDAY
        {'D': {1: CHOGHADIYU_LA,
               2: CHOGHADIYU_AM,
               3: CHOGHADIYU_KA,
               4: CHOGHADIYU_SH,
               5: CHOGHADIYU_RO,
               6: CHOGHADIYU_UD,
               7: CHOGHADIYU_CH,
               8: CHOGHADIYU_LA},
         'N': {1: CHOGHADIYU_UD,
               2: CHOGHADIYU_SH,
               3: CHOGHADIYU_AM,
               4: CHOGHADIYU_CH,
               5: CHOGHADIYU_RO,
               6: CHOGHADIYU_KA,
               7: CHOGHADIYU_LA,
               8: CHOGHADIYU_UD}
         },
    3:  # THURSDAY
        {'D': {1: CHOGHADIYU_SH,
               2: CHOGHADIYU_RO,
               3: CHOGHADIYU_UD,
               4: CHOGHADIYU_CH,
               5: CHOGHADIYU_LA,
               6: CHOGHADIYU_AM,
               7: CHOGHADIYU_KA,
               8: CHOGHADIYU_SH},
         'N': {1: CHOGHADIYU_AM,
               2: CHOGHADIYU_CH,
               3: CHOGHADIYU_RO,
               4: CHOGHADIYU_KA,
               5: CHOGHADIYU_LA,
               6: CHOGHADIYU_UD,
               7: CHOGHADIYU_SH,
               8: CHOGHADIYU_AM}
         },
    4:  # FRIDAY
        {'D': {1: CHOGHADIYU_CH,
               2: CHOGHADIYU_LA,
               3: CHOGHADIYU_AM,
               4: CHOGHADIYU_KA,
               5: CHOGHADIYU_SH,
               6: CHOGHADIYU_RO,
               7: CHOGHADIYU_UD,
               8: CHOGHADIYU_CH},
         'N': {1: CHOGHADIYU_RO,
               2: CHOGHADIYU_KA,
               3: CHOGHADIYU_LA,
               4: CHOGHADIYU_UD,
               5: CHOGHADIYU_SH,
               6: CHOGHADIYU_AM,
               7: CHOGHADIYU_CH,
               8: CHOGHADIYU_RO}
         },
    5:  # SATURDAY
        {'D': {1: CHOGHADIYU_KA,
               2: CHOGHADIYU_SH,
               3: CHOGHADIYU_RO,
               4: CHOGHADIYU_UD,
               5: CHOGHADIYU_CH,
               6: CHOGHADIYU_LA,
               7: CHOGHADIYU_AM,
               8: CHOGHADIYU_KA},
         'N': {1: CHOGHADIYU_LA,
               2: CHOGHADIYU_UD,
               3: CHOGHADIYU_SH,
               4: CHOGHADIYU_AM,
               5: CHOGHADIYU_CH,
               6: CHOGHADIYU_RO,
               7: CHOGHADIYU_KA,
               8: CHOGHADIYU_LA}
         },
    6:  # SUNDAY
        {'D': {1: CHOGHADIYU_UD,
               2: CHOGHADIYU_CH,
               3: CHOGHADIYU_LA,
               4: CHOGHADIYU_AM,
               5: CHOGHADIYU_KA,
               6: CHOGHADIYU_SH,
               7: CHOGHADIYU_RO,
               8: CHOGHADIYU_UD},
         'N': {1: CHOGHADIYU_SH,
               2: CHOGHADIYU_AM,
               3: CHOGHADIYU_CH,
               4: CHOGHADIYU_RO,
               5: CHOGHADIYU_KA,
               6: CHOGHADIYU_LA,
               7: CHOGHADIYU_UD,
               8: CHOGHADIYU_SH}
         }
}


def merge_date_and_time(give_dt, given_tm):
    return datetime(give_dt.year, give_dt.month, give_dt.day,
                    given_tm.hour, given_tm.minute, given_tm.second,
                    given_tm.microsecond)


def calculate(given_date, sunrise_time, sunset_time, next_sunrise_time):
    """
        This routine fills Choghadiya slots of CHOGHADIYA_SLOTS dictionary by
        first arriving at length of day and nigh choghadiya and then calculates
        all the choghadiya slots.
    """
    if not isinstance(given_date, date):
        print("Invalid data received as sunrise datetime.", file=sys.stderr)
        sys.exit(1)

    if not isinstance(sunrise_time, time):
        print("Invalid time received as sunrise time.", file=sys.stderr)
        sys.exit(1)

    if not isinstance(sunset_time, time):
        print("Invalid data received as sunset datetime.", file=sys.stderr)
        sys.exit(1)

    if not isinstance(next_sunrise_time, time):
        print("Invalid data received as next sunrise datetime.",
              file=sys.stderr)
        sys.exit(1)

    d1 = merge_date_and_time(given_date, sunrise_time)
    d2 = merge_date_and_time(given_date, sunset_time)
    d3 = merge_date_and_time((given_date + timedelta(days=1)), next_sunrise_time)

    if not (d1 < d2 < d3):
        print("Sunrise, sunset and next sunrise should be in chronological "
              "order.",
              file=sys.stderr)
        sys.exit(1)

    total_day_len = d3 - d1
    if total_day_len.days >= 2:
        err = "Can not calculate when next sunrise datetime " \
              "is more than one day apart.",
        raise ValueError(err)

    day_choghadiya_len = (d2 - d1) / 8
    night_choghadiya_len = (d3 - d2) / 8
    week_day_no = d1.weekday()

    #
    # This dictionary holds Choghadiya for a given day with their respective
    # start and end time.
    #
    slots = {
        'D':
            {
                1: {"N": "", "S": 0.0, "E": 0.0},
                2: {"N": "", "S": 0.0, "E": 0.0},
                3: {"N": "", "S": 0.0, "E": 0.0},
                4: {"N": "", "S": 0.0, "E": 0.0},
                5: {"N": "", "S": 0.0, "E": 0.0},
                6: {"N": "", "S": 0.0, "E": 0.0},
                7: {"N": "", "S": 0.0, "E": 0.0},
                8: {"N": "", "S": 0.0, "E": 0.0}
            },
        'N':
            {
                1: {"N": "", "S": 0.0, "E": 0.0},
                2: {"N": "", "S": 0.0, "E": 0.0},
                3: {"N": "", "S": 0.0, "E": 0.0},
                4: {"N": "", "S": 0.0, "E": 0.0},
                5: {"N": "", "S": 0.0, "E": 0.0},
                6: {"N": "", "S": 0.0, "E": 0.0},
                7: {"N": "", "S": 0.0, "E": 0.0},
                8: {"N": "", "S": 0.0, "E": 0.0}
            }
    }

    dst = d1
    den = dst + day_choghadiya_len
    nst = d2
    nen = nst + night_choghadiya_len

    for i in range(1, 9):
        slots['D'][i]["N"] = CHOGHADIYA_DICT[week_day_no]['D'][i]
        slots['N'][i]["N"] = CHOGHADIYA_DICT[week_day_no]['N'][i]
        slots['D'][i]["S"] = dst
        slots['N'][i]["S"] = nst
        slots['D'][i]["E"] = den
        slots['N'][i]["E"] = nen
        dst, den = den, den + day_choghadiya_len
        nst, nen = nen, nen + night_choghadiya_len
    return slots


def calculate_for_specific_time(given_date, sr_time, ss_time, nsr_time,
                                given_time):
    d1 = merge_date_and_time(given_date, sr_time)
    d3 = merge_date_and_time((given_date + timedelta(days=1)), nsr_time)
    d4 = merge_date_and_time(d3, time(0, 0, 0, 0))
    d5 = merge_date_and_time(given_date, given_time)
    if d1 <= d5 <= d4:
        pass
    else:
        d5 = d5 + timedelta(days=1)

    vaar = "{:%a}".format(given_date)
    x = calculate(given_date, sr_time, ss_time, nsr_time)
    for key, val in x.items():
        for subkey, subval in val.items():
            if subval['S'] <= d5 <= subval['E']:
                result = vaar, key, subkey, subval['N'], subval['S'], subval['E']
                return result
    return None


def main():
    message = """
    SIMPLE CHOGHADIYA CALCULATOR:-
    ============================
    Correct way to use this script is to import this script in your python 
    script and then call it as follows, passing three required arguments  
    
    >>> import scc
    >>> scc.calculate( d1, d2, d3)
    This will return entire dictionary of the all the choghadiya for given time.
    
    Where d1, d2 and d3 must be valid instances of python datetime class and 
    they must contain following values:
    
        d1 - must have value for Sunrise's date & time
        d2 - must have value for Sunset's date & time
        d3 - must have value for next day's Sunrise's date & time
        
    or
    if you need choghadiya information for just a point of time use following 
    method.
    
    >>> import scc
    >>> d1 = dttm.now().replace( hour=6, minute=30)
    >>> d2 = dttm.now().replace( hour=18, minute=30)
    >>> d3 = dttm.now().replace( day=17, hour=6, minute=30)
    >>> d4 = dttm.now()
    >>> scc.calculate_for_specific_time(d1, d2, d3, d4)
    
    This will give one single choghadiya for given time provided d1 <= d4 <= d3.
    """

    print(message)


if __name__ == '__main__':
    main()


src/vau.py
#!/usr/bin/env python3
# -*- coding: utf_8 -*-
"""

    Copyright    : 2015 May. A. R. Bhatt.
    Organization : VAU SoftTech
    Project      : scc
    Script Name  : vau.py
    License      : GNU General Public License v3.0


    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""
from pathlib import Path
from datetime import datetime as dttm
import argparse as ap
"""
    In reference to this article, here mimicking Java is avoided 
    by avoiding creation of "Singleton Class". 
    
    Because "Python is not Java"
    That insightful article is here... 
    https://dirtsimple.org/2004/12/python-is-not-java.html
"""


################################################################################
#
# FOLDER MANAGEMENT ROUTINES
#
################################################################################


def run_from_folder(script_name):
    if isinstance(script_name, str):
        script_name = Path(script_name)
    return script_name.parent


def get_user_home():
    return Path.home()


def get_cwd():
    return Path.cwd()


def is_running_from_home_folder():
    return get_user_home() == get_cwd()


def is_running_from_script_folder(script_name):
    script_folder = get_script_filepath(script_name)
    return script_folder == get_cwd()


def is_running_from_subfolder_of_cwd(script_name):
    x = get_script_filepath(script_name)
    y = get_cwd()
    return str(x).startswith(str(y)), x.relative_to(y)


def get_script_filename(script_path_and_name=None):
    if script_path_and_name is not None:
        result = Path(script_path_and_name).resolve().name
    else:
        result = get_script_filepath(__file__)
    return result


def get_script_filepath(script_path_and_name=None):
    if script_path_and_name is not None and \
            script_path_and_name != "":
        x = Path(script_path_and_name)
        if x.exists() and x.is_file():
            result = Path(script_path_and_name).resolve().parent
        elif x.exists() and x.is_dir():
            result = x.resolve()
        else:
            result = x
    else:
        result = get_script_filepath(__file__)
    return result


def folder_tree(folder_name):
    if isinstance(folder_name, str):
        folder_name = Path(folder_name)

    print(f'+ {folder_name}')

    for a_path in sorted(folder_name.rglob('*')):
        depth = len(a_path.relative_to(folder_name).parts)
        spacer = '---' * depth
        spacer = '+' + spacer
        print(f'{spacer}+ {a_path.name}')


def running_from_scripts_own_folder(script_name):
    if isinstance(script_name, str):
        script_name = Path(script_name)
    script_name = script_name.resolve()
    if script_name.is_file():
        return script_name.parent == get_cwd()
    elif script_name.is_dir():
        return script_name == get_cwd()


def get_absolute_path_to_subfolder(script_path, subfolder_name):
    return script_path / subfolder_name


################################################################################
#
# COMMAND LINE ARGUMENTS VALIDATOR
#
################################################################################
def date_type_validation(argument):
    try:
        return dttm.strptime(argument, "%Y-%m-%d")
    except ValueError:
        raise ap.ArgumentTypeError("Not a valid date: '{0}'.".format(argument))


def time_type_validation(argument):
    try:
        return dttm.strptime(argument, "%H:%M:%S")
    except ValueError:
        try:
            return dttm.strptime(argument, "%H:%M")
        except ValueError:
            raise ap.ArgumentTypeError("Not a valid time: '{0}'.".format(argument))


def main():
    txt = """
    VAU.PY by: VAU SOFTTECH 
    Author: A. R. Bhatt
    ======
    Misc Utility Script.
    This script is best used by importing it in to your code. 
    """
    print(txt)
    return


if __name__ == '__main__':
    main()


src/info.py
#!/usr/bin/env python3
# -*- coding: utf_8 -*-
"""

    Copyright    : 2015 May. A. R. Bhatt.
    Organization : VAU SoftTech
    Project      : SCC
    Script Name  : info.py
    License      : GNU General Public License v3.0


    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""


class Info:
    """
    This class manages basic Choghadiya information.

    Choghadiya Name : English Transliterated name of the Choghadiyu.
    Gujarati Name : Unicode Gujarati name of the Choghadiyu.
    Choghadiya Type : Auspicious, Inauspicious, or Neutral.
    Ruling Planet : One of the celestial bodies from the Sun, the Moon,
                    Mercury, Venus, Jupiter, Saturn etc...
    Good For :  What kinds of tasks can be undertaken : Mentions the
                types of tasks that can be undertaken when a particular
                choghadiyu is running.
    """

    """Some Constants declarations for Choghadiya names"""
    CH_UD = "Udveg"
    CH_CH = "Char"
    CH_LB = "Laabh"
    CH_AM = "Amrut"
    CH_KL = "Kal"
    CH_SB = "Shubh"
    CH_RG = "Rog"

    info_dict = {
        CH_UD: {
            "gu_nm": "ઉદ્વેગ",
            "type": "Inauspicious",
            "ruler": "Sun",
            "goodfor": "government related work"
        },
        CH_CH: {
            "gu_nm": "ચર",
            "type": "Neutral",
            "ruler": "Venus",
            "goodfor": "travel related work"
        },
        CH_LB: {
            "gu_nm": "લાભ",
            "type": "Auspicious",
            "ruler": "Mercury",
            "goodfor": "all kinds of gain related work"
        },
        CH_AM: {
            "gu_nm": "અમૃત",
            "type": "Auspicious",
            "ruler": "Moon",
            "goodfor": "all kinds of work"
        },
        CH_KL: {
            "gu_nm": "કાળ",
            "type": "Inauspicious",
            "ruler": "Saturn",
            "goodfor": "wealth accumulation"
        },
        CH_SB: {
            "gu_nm": "શુભ",
            "type": "Auspicious",
            "ruler": "Jupiter",
            "goodfor": "marriage & association"
        },
        CH_RG: {
            "gu_nm": "રોગ",
            "type": "Inauspicious",
            "ruler": "Mars",
            "goodfor": "defeat enemy"
        }
    }

    def __new__(cls):
        """
        We do not want to allow creation of the instances of this class. but it
        we do not mind if someone later on inherits from this class and creates
        instances of that  new class.
        :param :
        """
        if cls is Info:
            print(f"Creating an instance of the class {cls} is not allowed.")
            return None
        return object.__new__(cls)

    @staticmethod
    def _get_list():
        # AVOID CALLING THIS METHOD FROM ANYWHERE OUT OF THIS UNIT,
        # Having discovered that there are only seven Choghadiya and there is a
        # certain patten associated with  their occurrence, here we are doing
        # away with that large dictionary in scc.py that holds all the
        # Choghadiya for whole week.
        return list(Info.info_dict.keys()) * 5

    @staticmethod
    def get_ruling_planet(ch_nm):
        if ch_nm in Info.info_dict.keys():
            return Info.info_dict[ch_nm]["ruler"]
        else:
            return "Unknown planet"

    @staticmethod
    def get_type(ch_nm):
        """
        :param ch_nm: str - A valid choghadiyu name
        :return: str
        """
        if ch_nm in Info.info_dict.keys():
            return Info.info_dict[ch_nm]["type"]
        else:
            return "Unknown type"

    @staticmethod
    def get_good_for(ch_nm):
        """
        A static method that returns a string describing what that
        choghadiyu is good for.

        :param ch_nm: name of a Choghadiyu
        :return:
        """
        if ch_nm in Info.info_dict.keys():
            return Info.info_dict[ch_nm]["goodfor"]
        else:
            return "Nothing!"

    @staticmethod
    def get_gujarati_name(ch_nm):
        """
        A static method that returns a string describing Gujarati Language
        name of the choghadiyu.

        :param ch_nm: name of a Choghadiyu
        :return:
        """
        return Info.info_dict[ch_nm]["gu_nm"]

    @staticmethod
    def choghadiya_names_gen(weekday_number):
        """
        Earlier, we had a dictionary to hold hard-coded list of Choghadiya. but
        there is a pattern in occurrence of Choghadiya therefore it can be
        replaced with a generator.

        This function returns a generator object generates a full list of
        all the Choghadiya for a day passed to this function as an argument.

        :param weekday_number: iso week day number is expected
        :return: A generator object if argument provided is a valid iso week
            day number or a None.
        """
        if 0 <= weekday_number <= 7:
            lst = Info._get_list()
            offsets = (0, 3, 6, 2, 5, 1, 4, 0)
            this_offset = offsets[weekday_number]
            this_range = lst[this_offset:this_offset + 8] + \
                lst[this_offset + 5:this_offset + 21:2][::-1]
            for ch in this_range:
                yield ch
        else:
            return None


def main():
    from datetime import datetime as dttm
    xdt = dttm.now()
    idx = dttm.isoweekday(xdt)
    print("-"*61)
    print("Choghadiya for the week day number: {:} ({:%A})".format(idx, xdt))
    print("-"*61)
    x = list(Info.choghadiya_names_gen(idx))
    for i in range(0, 16):
        nm = x[i]
        good_for = (Info.get_good_for(nm)).capitalize()
        gj_nm = Info.get_gujarati_name(nm)
        print(f"{i + 1:>02}). {nm} \t{good_for:30.30s} \t{gj_nm}")
    return


if __name__ == '__main__':
    main()


src/scc-gui.py
#!/usr/bin/env python3
# -*- coding: utf_8 -*-
"""

    Copyright    : 2015 May. A. R. Bhatt.
    Organization : VAU SoftTech
    Project      : SCC
    Script Name  : scc-gui.py
    License      : GNU General Public License v3.0


    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""
import os
import argparse as ap
from pathlib import Path
from datetime import datetime as dttm, time as tm

from tkinter import *
from tkinter.ttk import *
from tkinter import messagebox as mb

from sccdateentry import DateEntry
from scctimeentry import TimeEntry

import vau
import choghadiya as ch


class CSSApp(Frame):

    def __init__(self, parent=None, app_title=None, app_icon=None, app_height=-1, app_width=-1):

        def create_sub_frames(container):
            container.result_win = None
            return

        def create_widgets(container):
            container.de = DateEntry(container, " Date : ")
            container.de.hide_hints()
            container.de.set(date_value=dttm.now())
            container.de.grid(row=0, column=0, padx=(10, 5), pady=(5, 0),
                              sticky=(W, E))

            container.sr = TimeEntry(container, " Sunrise Time :")
            container.sr.hide_hints()
            container.sr.set(time_value=tm(hour=6, minute=30, second=0))
            container.sr.grid(row=1, column=0, padx=(10, 5), pady=(0, 0),
                              sticky=(W, E))

            container.ss = TimeEntry(container, " Sunset Time :")
            container.ss.hide_hints()
            container.ss.set(time_value=tm(hour=18, minute=30, second=0))
            container.ss.grid(row=2, column=0, padx=(10, 5), pady=(0, 0),
                              sticky=(W, E))

            container.nsr = TimeEntry(container, " Next Sunrise Time :")
            container.nsr.hide_hints()
            container.nsr.set(time_value=tm(hour=6, minute=30, second=0))
            container.nsr.grid(row=3, column=0, padx=(10, 5), pady=(0, 0),
                               sticky=(W, E))

            container.cf = TimeEntry(container, " Calculate for time :")
            container.cf.hide_hints()
            container.cf.set(time_value=dttm.now())
            container.cf.grid(row=4, column=0, padx=(10, 5), pady=(0, 5),
                              sticky=(W, E))

            container.calc_full_btn = Button(container,
                                             text=" Calculate Choghadiya for full day",
                                             command=container.calculate_full)
            container.calc_full_btn.grid(row=0, column=1, rowspan=4,
                                         padx=(0, 10), pady=(0, 0),
                                         sticky=W + E + S + N)

            container.calc_btn = Button(container,
                                        text=" Calculate Choghadiya for this moment",
                                        command=container.calculate)
            container.calc_btn.grid(row=4, column=1, padx=(0, 10), pady=(0, 5),
                                    sticky=W + E + S + N)
            return

        if parent is not None:
            self.parent = parent
        else:
            self.parent = Tk()

        super(CSSApp, self).__init__(parent)
        if app_title is None:
            self.master.title(f"Simple Choghadiya Calculator by VAU")
        else:
            self.master.title(f"{app_title} by VAU")

        create_sub_frames(self)
        create_widgets(self)

        if app_icon is None:
            spn = vau.get_script_filepath(__file__)
            logo_file_name = spn / Path('config/logo.png')
            if os.path.exists(logo_file_name):
                photo = PhotoImage(file=logo_file_name)
                self.master.iconphoto(False, photo)
            else:
                print('logo file not found.')
        else:
            if os.path.exists(app_icon):
                photo = PhotoImage(file=app_icon)
                self.master.iconphoto(False, photo)
            else:
                print('{} file not found.'.format(app_icon))
        self.pack()
        self.master.protocol("WM_DELETE_WINDOW", self.client_exit)
        return

    def client_exit(self):
        if self.ask("Decide", "Are you sure to exit?") == 'yes':
            self.master.destroy()
        else:
            self.show_message('Return', 'Ok then, you will now return to the application screen.')
        return

    def run(self):
        self.parent.mainloop()
        return

    def ask(self, title, question):
        self.bell()
        return mb.askquestion(title, question)

    def show_message(self, title, message):
        self.bell()
        mb.showinfo(title, message)
        return

    def disable_all_btn(self):
        self.calc_btn.config(state='disabled')
        self.calc_full_btn.config(state='disabled')
        return

    def enable_all_btn(self):
        self.calc_btn.config(state='normal')
        self.calc_full_btn.config(state='normal')
        return

    def calculate_full(self):
        d1 = self.de.get
        d2 = self.sr.get
        d3 = self.ss.get
        d4 = self.nsr.get
        # answer = scc.calculate(d1, d2, d3, d4)
        answer = ch.Choghadiya(d1, d2, d3, d4)
        self.bell()
        self.show_choghadiya(answer)
        return

    def calculate(self):
        self.disable_all_btn()
        d1 = self.de.get
        d2 = self.sr.get
        d3 = self.ss.get
        d4 = self.nsr.get
        d5 = self.cf.get

        # answer = scc.calculate_for_specific_time(d1, d2, d3, d4, d5)
        answer = ch.Choghadiya(d1, d2, d3, d4)
        xch = answer.current_choghadiyu(d5)

        text_ans = """Currently, at {:%X}, on {:} {:} Choghadiyu is running. 
It started at {:%X} and will last till {:%X}.
        """.format(d5, answer.vedic_weekday_name, xch.ch_name,
                   xch.start, xch.end)
        self.bell()
        self.show_message("Result ...", text_ans)
        self.cf.set(dttm.now())
        self.enable_all_btn()
        return

    def show_choghadiya(self, xch):
        self.disable_all_btn()
        self.result_win = Toplevel(self.master)
        self.result_win.protocol("WM_DELETE_WINDOW", self.close_result_win)

        ttl = "Choghadiya for Vedic day \"{:%A}\"".format(self.de.get)
        self.result_win.title(ttl)

        day_title = Label(self.result_win, text="Day Choghadiya")
        day_title.grid(row=0, column=0, columnspan=4)
        night_title = Label(self.result_win, text="Night Choghadiya")
        night_title.grid(row=0, column=6, columnspan=4)

        close_btn = Button(self.result_win, text="Close this window",
                           command=self.close_result_win)
        close_btn.grid(row=10, column=0, columnspan=10, sticky=(W, E))

        for x in range(1, 9):
            l1 = Label(self.result_win, text="{:>02}".format(x))
            l1.grid(row=x, column=0, padx=(5, 0), pady=(5, 5), sticky=E)

            l2 = Label(self.result_win, text=xch[x]['name'])
            l2.grid(row=x, column=1, padx=(5, 5), pady=0, sticky=W)

            l3 = Label(self.result_win, text="{:%X}".format(xch[x]['start']))
            l3.grid(row=x, column=2, padx=(5, 5), pady=0, sticky=E)

            l4 = Label(self.result_win, text="{:%X}".format(xch[x]['end']))
            l4.grid(row=x, column=3, padx=(5, 15), pady=0, sticky=W)

        for x in range(9, 17):
            l1 = Label(self.result_win, text="{:>02}".format(x))
            l1.grid(row=x - 8, column=6, padx=(5, 0), pady=(5, 5), sticky=E)

            l2 = Label(self.result_win, text=xch[x]['name'])
            l2.grid(row=x - 8, column=7, padx=(5, 5), pady=0, sticky=W)

            l3 = Label(self.result_win, text="{:%X}".format(xch[x]['start']))
            l3.grid(row=x - 8, column=8, padx=(5, 5), pady=0, sticky=E)

            l4 = Label(self.result_win, text="{:%X}".format(xch[x]['end']))
            l4.grid(row=x - 8, column=9, padx=(5, 15), pady=0, sticky=W)

        vertical_sep = Separator(self.result_win, orient=VERTICAL)
        vertical_sep.grid(row=0, column=5, rowspan=10, sticky=(N, S))

        self.result_win.mainloop()
        return

    def close_result_win(self):
        self.enable_all_btn()
        self.result_win.destroy()
        return

    def set_date(self, new_value):
        if new_value is not None:
            self.de.set(new_value)
        return

    def set_sunrise(self, new_value):
        if new_value is not None:
            self.sr.set(new_value)
        return

    def set_sunset(self, new_value):
        if new_value is not None:
            self.ss.set(new_value)
        return

    def set_next_sunrise(self, new_value):
        if new_value is not None:
            self.nsr.set(new_value)
        return

    def set_calc_at(self, new_value):
        if new_value is not None:
            self.cf.set(new_value)


def main(arg1=None, arg2=None, arg3=None, arg4=None, arg5=None):
    app = CSSApp()
    app.set_date(arg1)
    app.set_sunrise(arg2)
    app.set_sunset(arg3)
    app.set_next_sunrise(arg4)
    app.set_calc_at(arg5)
    app.run()
    return


if __name__ == '__main__':
    scc_gui_parser = ap.ArgumentParser(prog='SCC',
                                       usage='%(prog)s [options]',
                                       description='Simple Choghadiya Calculator.',
                                       epilog="Simple and easy way to calculate "
                                              "Choghadiya that is part of Sanaatan "
                                              "Panchang."
                                       )
    scc_gui_parser.add_argument('-v', '--version', action='version',
                                version='%(prog)s 0.0.1 (Unstable Alpha)')

    date_group = scc_gui_parser.add_argument_group('date_group', 'Date Group')
    date_group.add_argument('-d', "--date",
                            help="Date - format YYYY-MM-DD",
                            required=False,
                            type=vau.date_type_validation,
                            default="{:%Y-%m-%d}".format(dttm.today()))

    events_group = scc_gui_parser.add_argument_group('events_group',
                                                     'Sun related events Group')
    events_group.add_argument("--sunrise",
                              help="Sunrise Time - format HH:MM[:SS]",
                              required=False,
                              type=vau.time_type_validation)
    events_group.add_argument("--sunset",
                              help="Sunset Time - format HH:MM[:SS]",
                              required=False,
                              type=vau.time_type_validation)
    events_group.add_argument("--next-sunrise",
                              help="Next Sunrise Time - format HH:MM[:SS]",
                              required=False,
                              type=vau.time_type_validation)
    scc_gui_parser.add_argument("--calc-at",
                                help="Calc for Time - format HH:MM[:SS]",
                                required=False,
                                type=vau.time_type_validation,
                                default="{:%H:%M:%S}".format(dttm.now().time()))
    args = scc_gui_parser.parse_args()

    if args.next_sunrise is None:
        args.next_sunrise = args.sunrise

    if args.calc_at is None:
        args.calc_at = dttm.now().time()
    main(args.date, args.sunrise, args.sunset, args.next_sunrise, args.calc_at)
